### ISSUE TITLE
🔴 Transactions are not used now!

### ISSUE THREAD

**Issue by Olegt0rr** [art049/odmantic#86]
# Bug

In the `_save()` method transaction's `session` is not passed to `update_one()` request. 
https://github.com/art049/odmantic/blob/541d4119fe5e558491ec21b223eb9242ccde4b91/odmantic/engine.py#L315-L320

**It means that transaction is NOT USED NOW!**

## How to fix?
It's not enough to just pass the `session` to request.
In MongoDB transactions are not available for Standalone. 
This simple way Standalone users will get exceptions for every `save()` call :cry: 

So we need to define the users of Standalone version and not start a transaction for them.
Example:
```python
if self.client._server_property('server_type') == SERVER_TYPE.Standalone:
    await self._save(instance, None, **kwargs)
else:
    async with await self.client.start_session() as s:
        async with s.start_transaction():
            await self._save(instance, s, **kwargs)
```

P.S.: I don't know how to properly check `ServerType` without protected `_server_property` method :(

**Comment by art049** [comment]
Nice catch @Olegt0rr :nerd_face:

> In MongoDB transactions are not available for Standalone.

Do you have some references about this ?
The only transaction requirements I found were those https://docs.mongodb.com/manual/core/write-operations-atomicity/

**Comment by art049** [comment]
To find out if we're connected to a replicaset or not, i found another way, using the [db.serverStatus](https://docs.mongodb.com/manual/reference/command/serverStatus/#serverstatus.repl) command:

```python
server_status = await self.database.command({'serverStatus'  :1})
is_standalone = "repl" not in server_status
```

I tried to use the `_server_property` you describe but I couldn't find it on my local client.
It would be probably faster to use it as it would avoid to send an extra request to the database.

**Comment by Olegt0rr** [comment]
> Do you have some references about this ?

https://docs.mongodb.com/manual/core/transactions-production-consideration/#feature-compatibility

Replica Set (4.0+) and Sharded Cluster (4.2+) supports transactions.
No Standalone listed here.
Also you can run experiment: pass session to request and your PR tester will answer with error. 


>  i found another way, using the [db.serverStatus](https://docs.mongodb.com/manual/reference/command/serverStatus/#serverstatus.repl) command

This way requires server admin role:
```not authorized on admin to execute command```

For my solution `readWrite` role is enough.

**Comment by art049** [comment]
>This way requires server admin role:
`not authorized on admin to execute command`

Yes :sob:, I totally forgot this initial constraint.

Thanks for creating the request in Jira !
I just saw an anwser there:
>**Bernie Hackett added a comment - Dec 25 2020 03:30:26 PM GMT+0000**
The easiest way to do what you want, with no API changes, is to call the ismaster command. If the msg field is set to "isdbgrid" you're connected to a mongos. If the setname field is set you're connected to a replica set memeber. Otherwise you're connected to a standalone.

This looks like a very good solution. We will have to check the version as well to make sure sharded cluster are 4.2+.
To add these commands in the Engine initializer, we might have to gather the event loop held by the motor client or maybe access the underlying pymongo client (I'm not sure it's doable in a clean way for this one)

**Comment by Olegt0rr** [comment]
Seems calling `self.client._server_property('server_type')` still easiest way 😭

**Comment by Olegt0rr** [comment]
Need some help to finish tests coverage

**Comment by art049** [comment]
Hey, yes I will merge #91 as soon as possible, this should help you with the coverage :)

**Comment by art049** [comment]
Closed by #227
